Stats Improvements Step 1: Added UI to support the new stats v4 api

anitaa1990 opened the pull request:
Fixes #1259. This PR completes the first step of the Stats Improvements Project as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1199). 

### Changes:
- [x] Adds a new FluxC hash that includes the latest v4 api changes.
- [x] Checks if v4 stats api is supported:
  - [x] Adds a new `RevenueStatsAvailabilityFetcher` class that handles the logic to fetch the stats v4 api availability for a site and stores the availability flag in `SharedPreferences`.
  - [x] The availability check is added to the `MainActivity` class and is called every time the app is opened.
  - [x] The availability check is added to the `SitePickerActivity` class and is called every time the user switches a store or picks a store after logging to the app.
  - [x] The availability check is added to the pull to refresh gesture in the old and the new stats. 
- [x] Duplicates the current Dashboard architecture for displaying the new Stats data - This include s duplicating the existing implementation of `DashboardFragment`, `DashboardPresenter`, `DashboardContract`, `DashboardStatsView`, `DashboardModule`. The new classes are added to a separate package called `mystore`.
- [x] The TAB bar title for the v4 stats UI has been changed to `Today`, `This Week`, `This Month`, and `This Year`.
- [x] Since the date ranges have been modified, the date range is passed to FluxC when querying for the visitor stats in v4 UI. 
- [x] Remove the top selling products tab layout.
- [x] Anchor the tabLayout when the view is scrolled.

 ### Notes:
- There is no logic in place, at the moment, to change the UI, if it is already visible to the user. This will be handled in a separate PR. For instance:
  - If the api response returned is successful but the user has already seen the v3 stats screen.
  - If the api response returned is error but the user has already seen the v4 stats screen (happens when the user has disabled this plugin). 
- The v4 stats api does not provide the currency for the data in the api response. So, the logic is modified to fetch the currency from the `Site Settings`.

### Behaviour:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62352407-b2d17680-b525-11e9-8f70-ff0a3295d2b4.gif">

### Screenshots:
#### No Data UI
(LEFT: old stats data . RIGHT: v4 stats data):
<img width="300" src="https://user-images.githubusercontent.com/22608780/62279898-96233900-b468-11e9-8bfa-abfe3e7ec86e.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62352454-d4caf900-b525-11e9-85a3-ffbf90a2ca38.png"> 

#### V4 Stats UI
<img width="300" src="https://user-images.githubusercontent.com/22608780/62352527-fc21c600-b525-11e9-8b6b-aaa053682be6.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62352531-ff1cb680-b525-11e9-8716-a920aa9aa04b.png"> 
<img width="300" src="https://user-images.githubusercontent.com/22608780/62352574-1a87c180-b526-11e9-8525-a94452aa1c3c.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62352578-1cea1b80-b526-11e9-8cab-7a5353b79614.png"> 

### Testing
- Run `MyStorePresenterTest`.
- Run `DateUtilsTest`
- Test a new order completion and verify that the Stats is updated for the current day.
- Test on site without `wc-admin` installed/activated.
- Test on site with `wc-admin` installed.
- Test on site without `wc-admin`, then install `wc-admin` and try again for the same site. The app will not refresh automatically. We would need to close the app and open again to verify the changes.

### Questions:
~~I accidentally changed the style of `TabLayout` so now there is an elevation added to the old stats UI as well. I thought it looked good so I left it that way 😊. We could certainly remove it without much hassle. But WDYT?~~ Removing the elevation for the old stats since there are two tabs and it looks confusing.
~~I noticed while testing that since today is the first of August and there are no orders for my test site today, that:~~
  - ~~Today TAB: has 0 stats.~~
  - ~~This week TAB: has some data (since the stats data is from the start of this week i.e. 29 July).~~
  - ~~This Month TAB: has no data (since the stats data is from the start of the current month i.e. today).~~
  - ~~This Year TAB: has data (since the stats data is from the start of this year)~~

~~I'm not sure if the difference between the data in WEEK and MONTH tab, would be confusing to the user. 😕~~ As @nbradbury mentioned [here](https://github.com/woocommerce/woocommerce-android/pull/1314#issuecomment-517279821), this might not be a problem since we would be adding the date bar to the tabs in another PR.

anitaa1990 committed:
Added latest fluxC hash

anitaa1990 committed:
Added method to parse date string + hour to readable hour format

anitaa1990 committed:
Added UI for the new v4 stats data

anitaa1990 committed:
Added unit tests for MyStorePresenter class

anitaa1990 committed:
Added logic in `BottomNavigationPosition` to display store fragment based on the v4 stats availability

anitaa1990 committed:
Added logic to update the v4 revenue stats api availability when switching stores/logging in

anitaa1990 committed:
Added helper class to check for v4 revenue stats api availability

anitaa1990 committed:
Modified logic check for v4 revenue stats availability when switching switching stores/logging in

anitaa1990 committed:
Added logic check for v4 revenue stats availability when app is opened

anitaa1990 committed:
Added logic to check for v4 revenue stats availability when pull to refresh is initiated

anitaa1990 committed:
Reverting logic to check v4 availability when MyStoreFragment is refreshed

This seemed redundant since we would be fetching the v4 stats data anyway when pull to refresh is called. Also added logic to display an error message if we find that the v4 stats support is no longer available

anitaa1990 committed:
Added method to switch between the two different versions of stats

anitaa1990 committed:
Removed top earners tabLayout and modified logic to load top earners when stats tab is selected

anitaa1990 committed:
Moved the TabLayout to the fragment xml to support anchoring of TabLayout when scrolling

anitaa1990 committed:
Added logic to display empty view when there is no data for a particular data range

anitaa1990 committed:
Add date range to visitor stats query to FluxC to support v4 UI date ranges

nbradbury commented:
> * I accidentally changed the style of `TabLayout` so now there is an elevation added to the old stats UI as well.

I love the new date ranges but I'm not a fan of the elevation.

nbradbury commented:
> I'm not sure if the difference between the data in WEEK and MONTH tab, would be confusing to the user

I think it will be somewhat confusing but it's not a big deal, especially once we add the date bar to the top.

nbradbury commented:
The gap between the cards makes me think they're not connected by the same date range, but maybe that's because I'm used to two tabs?

![new stats](https://user-images.githubusercontent.com/3903757/62296484-2cda0080-b43d-11e9-9f0e-f3edad0257d5.png)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/RevenueStatsAvailabilityFetcher.kt:
```diff
@@ -0,0 +1,42 @@
+    @Suppress("unused")
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    fun onWCRevenueStatsChanged(event: OnWCRevenueStatsChanged) {
+        if (event.causeOfChange == FETCH_REVENUE_STATS_AVAILABILITY) {
```
I'm wondering if we need to do anything to handle when `event.isError` is true? Looking at [the FluxC code](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/dc9975cd0dc435cbd9fcab1f6366141535650599/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt#L633) I see we assume the v4 API isn't available when an error occurs, which is fine, but it does mean a user could have seen the new stats the previous time the app was launched but now sees the old stats due to an error.

anitaa1990 replied:
>  user could have seen the new stats the previous time the app was launched but now sees the old stats due to an error.

Great catch! I fixed this in fcaf9e0. While testing this scenario, I noticed that we receive the `GENERIC_ERROR` message when trying to check the availability when there is no internet. So I modified the logic to update `SharedPreferences` only if there is no error or if the error response is because of the the `wc-admin` plugin not being available.

nbradbury commented on WooCommerce/src/main/res/values/styles.xml:
```diff
@@ -309,9 +309,12 @@
-        <item name="tabPaddingEnd">40dp</item>
+        <item name="tabSelectedTextColor">@color/black</item>
+        <item name="android:background">@color/card_bgColor</item>
+        <item name="android:elevation">10dp</item>
```
If we use elevated tabs, let's store this value in `dimens.xml` with the other elevations. Also, `10dp` seems quite high - the appbar is only `4dp` and cards are only `1dp`.

anitaa1990 replied:
Fixed in 453b4f2. I modified the styles after looking through the designs and removed the margins from the stats card and the top earners card.

nbradbury commented on WooCommerce/src/main/res/values/styles.xml:
```diff
@@ -552,4 +555,13 @@
+         Stats
+        -->
+    <style name="Woo.Stats.Card" parent="Woo.Card">
+        <item name="android:elevation">2dp</item>
```
Note that we already have a `card_elevation` in `dimens.xml` which is `dp`.

anitaa1990 replied:
I couldn't see a visible elevation when testing on my emulator so I had increased it to `2dp`.  But changed this in after testing in 453b4f2.

<img width="300" src="https://user-images.githubusercontent.com/22608780/62353521-79e6d100-b528-11e9-9523-eb44b3915b73.png">

anitaa1990 committed:
Bug fix: The toolbar title not changed after opening a product detail and clicking back

anitaa1990 committed:
Modified styles/UI after checking with design changes

anitaa1990 committed:
Fix lint error

anitaa1990 committed:
Modified logic to store availability only if there is no error or if the error is because of plugin not being active

anitaa1990 committed:
If the v4 api is not available but the v4 UI is displayed, show an error snack and update the local cache

anitaa1990 commented:
Thanks for the quick review @nbradbury!!

I found a bug when testing where the toolbar title after opening a product detail from top earners, is not changed when back is clicked. Fixed this in bb13cd7. I also modified the styles after comparing them with the designs. So removed the margins from the Stats card and the Top earners card and fixed the elevation in v4 UI. 

> The gap between the cards makes me think they're not connected by the same date range, but maybe that's because I'm used to two tabs?

There was an increase in the gaps between the two cards. It was not included in the design. So I modified that as well in 453b4f2.

> I don't like the elevation when there are two tab layouts, but on sites with wc-admin which show only a single tab layout I do kind of like it since the tabs are fixed to the top. 

I agree that after testing on a device, the elevation on the current stats don't look good, seeing as there are two separate tabs available. So removed the elevation in the current stats. 

> I'm not sure how this would look, but I'm wondering if we should try removing the elevation and only add it when the dashboard is scrolled?

I totally forgot that we would be adding the date bar in a different PR 🤦‍♀ 

I've updated the latest screenshots to the PR description and marked the PR ready for review 👍

nbradbury commented:
> * Test on site without `wc-admin`, then install `wc-admin` and try again for the same site. The app will not refresh automatically. We would need to close the app and open again to verify the changes.

This concerned me at first, but it's not really a big deal since you're calling `fetchRevenueStatsAvailability` when the user does a pull-to-refresh.

nbradbury commented:
This all looks great @anitaa1990! I'm sad that `wc-admin` is delayed in core because these changes are such an improvement, but at least `wc-admin` beta testers will see them. :shipit:
